---
title: "GenBank_Data"
author: "Biancani"
date: "April 8, 2016"
output: pdf_document
---
#Download Data from GenBank:

##Entrez Direct (EDirect) Installation
http://www.ncbi.nlm.nih.gov/books/NBK179288/#chapter6.Getting_Started

At UNIX command line, enter:
```{bash, eval=FALSE}
cd ~
perl -MNet::FTP -e \
'$ftp = new Net::FTP("ftp.ncbi.nlm.nih.gov", Passive => 1); $ftp->login;
$ftp->binary; $ftp->get("/entrez/entrezdirect/edirect.zip");'
unzip -u -q edirect.zip
rm edirect.zip
export PATH=$PATH:$HOME/edirect
./edirect/setup.sh
```

##Use EDirect to Download Fasta & GenBank Files

At UNIX command line enter:
```{bash, eval=FALSE}
# specify options for fasta files:
#QUERY=scinidae; DATABASE=popset; FORMAT=fasta
#QUERY=hyperia; DATABASE=nucleotide; FORMAT=fasta
#QUERY=Hyperiidea; DATABASE=nucleotide; FORMAT=fasta
QUERY=Gammaridea; DATABASE=nucleotide; FORMAT=fasta


FILENAME=$DATABASE.$QUERY.$FORMAT
esearch -db $DATABASE -query $QUERY | efetch -format $FORMAT > $FILENAME

#options to change for genbank files:
FORMAT=gb

FILENAME=$DATABASE.$QUERY.$FORMAT
esearch -db $DATABASE -query $QUERY | efetch -format $FORMAT > $FILENAME
```

##Read Data into R

```{r}
#specify options used above for EDirect:
#QUERY="scinidae"
#QUERY="hyperia"
#QUERY="Hyperiidea"
QUERY="Gammaridea"


#DATABASE="popset"
DATABASE="nucleotide"

fasta_filename=paste(DATABASE,QUERY,"fasta",sep=".")
genbank_filename=paste(DATABASE,QUERY,"gb",sep=".")

#read data into R
fasta=readLines(fasta_filename)
genbank=readLines(genbank_filename)
```

##Parse Data

```{r, eval=FALSE, tidy=TRUE}
#load stringr library
library(stringr)

#create vector of genbank data
genbank_vector=c()
gb=""
for(line in genbank){
  gb=paste(gb,line,sep=" ")
  if(str_detect(line,"//")){
    gb=str_replace_all(gb," +"," ")
    genbank_vector=append(genbank_vector,gb)
    gb=""
  }
}

genelookup= data.frame(
  rawlist=c("histone H{0,1}3","12S","16S","18S","small subunit ribosomal","28S","cytochrome [c ]{0,2}oxidase [subunit ]{0,8}[I1] ","[cC][Oo][Xx]{0,1}[1Ii][ )]","RpS5","RpS2","ArgKin","EF1-{0,1}a","GAPDH","IDH","MDH","CAD","wingless","Unknown sequence","COX2","COX3","ATP6","cytochrome b","HSP70","mitochondrion, complete genome","internal transcribed spacer 2"),
  genelist=c("H3","12S","16S","18S","18S","28S","COI","COI","RpS5","RpS2","ArgKin","EF1a","GAPDH","IDH","MDH","CAD","Wgl","Unknown","COX2","COX3","ATP6","CytB","HSP70","complete mito genome","ITS2")
)

#initiate vectors
fasta_header=c()
sequence=c()
accession=c()
description=c()
organism=c()
voucher=c()
isolate=c()
gene=c()
bp_length=c()
taxon_id=c()
author=c()
author_ref2=c()

others=c()

#parse fasta (& genbank) data
seq=""
record=0
for (line in fasta) {
  if (str_detect(line,fixed(">"))){
    # if a header is detected:
    record=record+1 #track record number
    fasta_header = append(fasta_header,line)
    header_split=strsplit(line, "|", fixed=TRUE)
    accession=append(accession,header_split[[1]][4])
    des=str_trim(header_split[[1]][5], side="left")
    description=append(description,des)
    #[cfa. ]{0,5} added to accomodate "Genus cf. species" and "Genus aff. species"
    organism=append(organism,str_extract(des,"[A-Z][a-z]+ [cfa. ]{0,5}[a-z.]+"))
    ingenes=FALSE
    gen=""
    for(i in seq(length(genelookup$rawlist))){
      if(str_detect(des,as.character(genelookup$rawlist[i]))){
        lookup=as.character(genelookup$genelist[i])
        if(!str_detect(gen,lookup)){gen=paste(lookup,gen,sep=" ")}
        ingenes=TRUE
      }
    }
    if(!ingenes){
      gen = "other"
      others=append(others,des)
    }
    gene=append(gene,gen)
    
    #data from genbank file:
    gb=genbank_vector[record]
    
    #voucher number from gb
    vou_pattern="voucher=\"([^\"]+)\""
    vou=str_replace(str_extract(gb,vou_pattern),vou_pattern,"\\1")
    voucher=append(voucher,vou)
    
    #isolate number from gb
    #iso_pattern="/isolate=\"([\\w ./-]+)\""
    iso_pattern="/isolate=\"([^\"]+)\""
    iso=str_replace(str_extract(gb,iso_pattern),iso_pattern,"\\1")
    isolate=append(isolate,iso)
    
    #sequence length from gb
    bps=as.numeric(str_extract(str_extract(gb,"[:digit:]+ bp"),"[:digit:]+"))
    bp_length=append(bp_length,bps)
    
    #taxon ID from gb
    tax=as.numeric(str_extract(str_extract(gb,"taxon:[:digit:]+"),"[:digit:]+"))
    taxon_id=append(taxon_id,tax)
    #authors from gb
    ref1pattern="REFERENCE 1[^,]*AUTHORS ([^\\d]+) TITLE"
    auth= str_replace(str_extract(gb,ref1pattern),ref1pattern,"\\1")
    author=append(author,auth)
    ref2pattern="REFERENCE 2[^,]*AUTHORS ([^\\d]+) TITLE"
    auth2= str_replace(str_extract(gb,ref2pattern),ref2pattern,"\\1")
    author_ref2=append(author_ref2,auth2)
    #if sequence was recorded for previous record, add it to the sequence vector
    if (nchar(seq)>0) sequence=append(sequence,seq)
    seq="" #reset seq
  } else {
    # if the line is not a header, add the line to seq
    seq=paste(seq,line,sep="")
  }
}
sequence=append(sequence,seq)

#print vectors
#fasta_header
#sequence
#accession
#description
#organism
#voucher
isolate
#gene
#bp_length
#taxon_id
#author
#author_ref2
others


df=data.frame(organism,taxon_id,voucher,isolate,accession,gene,bp_length,author,author_ref2,description,fasta_header,sequence)

```
