---
title: "GenBank_Data"
author: "Biancani"
date: "April 8, 2016"
output: pdf_document
---
#Download Data from GenBank:

##Entrez Direct (EDirect) Installation
http://www.ncbi.nlm.nih.gov/books/NBK179288/#chapter6.Getting_Started

At UNIX command line, enter:
```{bash, eval=FALSE}
cd ~
perl -MNet::FTP -e \
'$ftp = new Net::FTP("ftp.ncbi.nlm.nih.gov", Passive => 1); $ftp->login;
$ftp->binary; $ftp->get("/entrez/entrezdirect/edirect.zip");'
unzip -u -q edirect.zip
rm edirect.zip
export PATH=$PATH:$HOME/edirect
./edirect/setup.sh
```

##Use EDirect to Download Fasta & GenBank Files

At UNIX command line enter:
```{bash, eval=FALSE}
# specify options for fasta files:
#QUERY=scinidae; DATABASE=popset; FORMAT=fasta
#QUERY=hyperia; DATABASE=nucleotide; FORMAT=fasta
#QUERY=Hyperiidea; DATABASE=nucleotide; FORMAT=fasta
QUERY=Gammaridea; DATABASE=nucleotide; FORMAT=fasta


FILENAME=$DATABASE.$QUERY.$FORMAT
esearch -db $DATABASE -query $QUERY | efetch -format $FORMAT > $FILENAME

#options to change for genbank files:
FORMAT=gb

FILENAME=$DATABASE.$QUERY.$FORMAT
esearch -db $DATABASE -query $QUERY | efetch -format $FORMAT > $FILENAME
```

##Read Data into R

```{r}
#specify options used above for EDirect:
#QUERY="scinidae"
#QUERY="hyperia"
#QUERY="Hyperiidea"
QUERY="Gammaridea"


#DATABASE="popset"
DATABASE="nucleotide"

fasta_filename=paste(DATABASE,QUERY,"fasta",sep=".")
genbank_filename=paste(DATABASE,QUERY,"gb",sep=".")

#read data into R
fasta=readLines(fasta_filename)
genbank=readLines(genbank_filename)
```

##Parse Data

```{r, eval=FALSE, tidy=TRUE}
#load stringr library
library(stringr)

#create vector of genbank data
genbank_vector=c()
gb=""
for(line in genbank){
  gb=paste(gb,line,sep=" ")
  if(str_detect(line,"//")){
    gb=str_replace_all(gb," +"," ")
    genbank_vector=append(genbank_vector,gb)
    gb=""
  }
}

genelookup= data.frame(
  rawlist=c("histone H{0,1}3","12S","16S","18S","small subunit ribosomal","28S","cytochrome [c ]{0,2}oxidase [subunit ]{0,8}[I1] ","[cC][Oo][Xx]{0,1}[1Ii][ )]","RpS5","RpS2","ArgKin","EF1-{0,1}a","GAPDH","IDH","MDH","CAD","wingless","Unknown sequence","COX2","COX3","ATP6","cytochrome b","HSP70","mitochondrion, complete genome","internal transcribed spacer 2"),
  genelist=c("H3","12S","16S","18S","18S","28S","COI","COI","RpS5","RpS2","ArgKin","EF1a","GAPDH","IDH","MDH","CAD","Wgl","Unknown","COX2","COX3","ATP6","CytB","HSP70","complete mito genome","ITS2")
)

#initiate vectors
fasta_header=c()
sequence=c()

species_name=c()
organism=c()
accession=c()
description=c()

voucher=c()
isolate=c()
gene=c()
bp_length=c()
taxon_id=c()

author_ref1=c()
title_ref1=c()
journal_ref1=c()

author_ref2=c()
title_ref2=c()
journal_ref2=c()

author_ref3=c()
title_ref3=c()
journal_ref3=c()

others=c()

#parse fasta (& genbank) data
seq=""
record=0
for (line in fasta) {
  # if a header is detected:
  if (str_detect(line,fixed(">"))){
    
    #if sequence was recorded for previous record, add it to the sequence vector
    if (nchar(seq)>0) sequence=append(sequence,seq)
    seq="" #reset seq
    
    record=record+1 #track record number
    
    #store fasta header
    fasta_header = append(fasta_header,line)
    
    #parse fasta header
    header_split=strsplit(line, "|", fixed=TRUE)
    
    #accession from fasta
    accession=append(accession,header_split[[1]][4])
    
    #description from fasta
    des=str_trim(header_split[[1]][5], side="left")
    description=append(description,des)
    
    #species name from fasta
    #[cfa. ]{0,5} added to accomodate "Genus cf. species" and "Genus aff. species"
    species_name=append(species_name,str_extract(des,"[A-Z][a-z]+ [cfa. ]{0,5}[a-z.]+"))
    
    #gene from fasta
    ingenes=FALSE
    gen=""
    for(i in seq(length(genelookup$rawlist))){
      if(str_detect(des,as.character(genelookup$rawlist[i]))){
        lookup=as.character(genelookup$genelist[i])
        if(!str_detect(gen,lookup)){gen=paste(lookup,gen,sep=" ")}
        ingenes=TRUE
      }
    }
    if(!ingenes){
      gen = "other"
      others=append(others,des)
    }
    gene=append(gene,gen)
    
    #parse data from genbank file:
    gb=genbank_vector[record]
    
    #organism from gb
    org_pattern="organism=\"([^\"]+)\""
    org=str_replace(str_extract(gb,org_pattern),org_pattern,"\\1")
    organism=append(organism,org)
    
    #voucher number from gb
    vou_pattern="voucher=\"([^\"]+)\""
    vou=str_replace(str_extract(gb,vou_pattern),vou_pattern,"\\1")
    voucher=append(voucher,vou)
    
    #isolate number from gb
    iso_pattern="/isolate=\"([^\"]+)\""
    iso=str_replace(str_extract(gb,iso_pattern),iso_pattern,"\\1")
    isolate=append(isolate,iso)
    
    #sequence length from gb
    bps=as.numeric(str_extract(str_extract(gb,"[:digit:]+ bp"),"[:digit:]+"))
    bp_length=append(bp_length,bps)
    
    #taxon ID from gb
    tax=as.numeric(str_extract(str_extract(gb,"taxon:[:digit:]+"),"[:digit:]+"))
    taxon_id=append(taxon_id,tax)
    
    #author and/or consortium for References 1-3 from gb
    reference_ends=c("REFERENCE","COMMENT","FEATURES")
    for(end in reference_ends){gb=str_replace_all(gb,end,paste("@",end,sep=""))}
    reference_keys=c("AUTHORS","CONSRTM","TITLE","JOURNAL")
    for(key in reference_keys){gb=str_replace_all(gb,key,paste("%",key,sep=""))}
    
    #REFERENCES 1-3
    for (i in 1:3) {
      auth_pat=paste("REFERENCE ","[^@]+%AUTHORS ([^%]+) %",sep=as.character(i))
      cons_pat=paste("REFERENCE ","[^@]+%CONSRTM ([^%]+) %",sep=as.character(i))
      title_pat=paste("REFERENCE ","[^@]+%TITLE ([^%]+) %",sep=as.character(i))
      journ_pat=paste("REFERENCE ","[^@]+%JOURNAL ([^%@]+) [%@]",sep=as.character(i))
      auth= str_replace(str_extract(gb,auth_pat),auth_pat,"\\1")
      cons= str_replace(str_extract(gb,cons_pat),cons_pat,"\\1")
      title= str_replace(str_extract(gb,title_pat),title_pat,"\\1")
      journ= str_replace(str_extract(gb,journ_pat),journ_pat,"\\1")
      if(is.na(auth)){
        auth=cons # in case consortium listed but no author
      }else if(!is.na(cons)){
        auth=paste(auth,cons,sep=", ") # in case author & consortium listed (never?)
      }
      if(i==1){
        author_ref1=append(author_ref1,auth)
        title_ref1=append(title_ref1,title)
        journal_ref1=append(journal_ref1,journ)
      }else if(i==2){
        author_ref2=append(author_ref2,auth)
        title_ref2=append(title_ref2,title)
        journal_ref2=append(journal_ref2,journ)
      }else if(i==3){
        author_ref3=append(author_ref3,auth)
        title_ref3=append(title_ref3,title)
        journal_ref3=append(journal_ref3,journ)
      }
    }
    
  } else {
    # if the line is not a header, add the line to seq
    seq=paste(seq,line,sep="")
  }
}
sequence=append(sequence,seq)

df=data.frame(species_name,organism,taxon_id,voucher,isolate,accession,gene,bp_length,author_ref1,title_ref1,journal_ref1,author_ref2,title_ref2,journal_ref2,author_ref3,title_ref3,journal_ref3,description,fasta_header,sequence)

print(others)

```
