---
title: "GenBank_Data"
author: "Biancani"
date: "April 8, 2016"
output: pdf_document
---
#Download Data from GenBank:

##Entrez Direct (EDirect) Installation
http://www.ncbi.nlm.nih.gov/books/NBK179288/#chapter6.Getting_Started

At UNIX command line, enter:
```{bash, eval=FALSE}
cd ~
perl -MNet::FTP -e \
'$ftp = new Net::FTP("ftp.ncbi.nlm.nih.gov", Passive => 1); $ftp->login;
$ftp->binary; $ftp->get("/entrez/entrezdirect/edirect.zip");'
unzip -u -q edirect.zip
rm edirect.zip
export PATH=$PATH:$HOME/edirect
./edirect/setup.sh
```

##Use EDirect to Download Fasta & GenBank Files

At UNIX command line enter:
```{bash, eval=FALSE}
# specify options for fasta files:
#TAXA=scinidae; DATABASE=popset; FORMAT=fasta
#TAXA=hyperia; DATABASE=nucleotide; FORMAT=fasta

#TAXA=Hyperiidea; DATABASE=nucleotide; FORMAT=fasta
#TAXA=Gammaridea; DATABASE=nucleotide; FORMAT=fasta
#TAXA=Ingolfiellidea; DATABASE=nucleotide; FORMAT=fasta
##TAXA=Senticaudata; DATABASE=nucleotide; FORMAT=fasta #too large?

##subgroups of Senticaudata:
#TAXA=Bogidiellida; DATABASE=nucleotide; FORMAT=fasta
#TAXA=Corophiida; DATABASE=nucleotide; FORMAT=fasta
##TAXA=Gammarida; DATABASE=nucleotide; FORMAT=fasta #causes errors #subgroup to find
#TAXA=Hadziida; DATABASE=nucleotide; FORMAT=fasta #downloaded, not run yet

#TAXA=Talitrida; DATABASE=nucleotide; FORMAT=fasta #not downloaded yet
#TAXA=unclassified Senticaudata

###subgroups of Gammarida:
#TAXA=Crangonyctidira; DATABASE=nucleotide; FORMAT=fasta #done no errors
##TAXA=Gammaridira; DATABASE=nucleotide; FORMAT=fasta #very large
###subgroups of Gammaridira:
#Acanthogammaridae
#Anisogammaridae
#Baikalogammaridae
#Eoniphargidae
#Eulimnogammaridae
#Gammarellidae
#Gammaridae
#Luciobliviidae
#Macrohectopidae
#Mesogammaridae
#Micruropodidae
#Pachyschesidae
#Pallaseidae
#Pontogammaridae
TAXA=Pontogammaridae; DATABASE=nucleotide; FORMAT=fasta
#Typhlogammaridae


QUERY=$TAXA[ORGN]
FILENAME=$DATABASE.$TAXA.$FORMAT
esearch -db $DATABASE -query $QUERY | efetch -format $FORMAT > $FILENAME

#options to change for genbank files:
FORMAT=gb

FILENAME=$DATABASE.$TAXA.$FORMAT
esearch -db $DATABASE -query $QUERY | efetch -format $FORMAT > $FILENAME

```

##Read Data into R

```{r}
#specify options used above for EDirect:
#TAXA="scinidae"
#TAXA="hyperia"

#TAXA="Hyperiidea"
#TAXA="Gammaridea"
#TAXA="Ingolfiellidea"
##TAXA="Senticaudata" #not downloaded - too large?
#TAXA="Bogidiellida"
#TAXA="Corophiida"

##TAXA="Gammarida" #error subgroups:
TAXA="Crangonyctidira"


#TAXA="Hadziida"
#TAXA="Talitrida"

#DATABASE="popset"
DATABASE="nucleotide"

fasta_filename=paste(DATABASE,TAXA,"fasta",sep=".")
genbank_filename=paste(DATABASE,TAXA,"gb",sep=".")

#read data into R
fasta=readLines(fasta_filename)
genbank=readLines(genbank_filename)
```

##Parse Data

```{r, eval=FALSE, tidy=TRUE}
#load stringr library
library(stringr)

#create vector of genbank data
genbank_vector=c()
gb=""
for(line in genbank){
  gb=paste(gb,line,sep=" ")
  if(str_detect(line,"//")){
    gb=str_replace_all(gb," +"," ")
    genbank_vector=append(genbank_vector,gb)
    gb=""
  }
}

genelookup= data.frame(
  rawlist=c("cytochrome [c ]{0,2}oxidase [subunit ]{0,8}[I1] ","[cC][Oo][Xx]{0,1}[1Ii][ )]","histone H{0,1}3","12S","16S","large subunit ribosomal RNA gene, partial sequence; mitochondrial","18S","small subunit ribosomal","28S","RpS5","RpS2","ArgKin","EF-{0,1}1-{0,1}a","GAPDH","IDH","MDH","CAD","wingless","COX2","COX3","ATP6","cytochrome b","HSP70","mitochondri[onal DNA]{2,6}, complete genome","complete mitochondrial genome","partial mitochondrial genome","internal transcribed spacer 2","internal transcribed spacer 1","transposon","Unknown sequence","anonymous locus","NAK-ATPase","sodium/potassium ATPase","prostaglandin E synthase","cyclooxygenase","GH7D","GH7A","superoxide dismutase gene","hemocyanin","CsDll","contig_\\d+_\\d+ transcribed RNA","microsatellite"),
  genelist=c("COI","COI","H3","12S","16S","16S","18S","18S","28S","RpS5","RpS2","ArgKin","EF1a","GAPDH","IDH","MDH","CAD","Wgl","COX2","COX3","ATP6","CytB","HSP70","mt Genome","mt Genome","partial mt Genome","ITS2","ITS1","transposon","Unknown","anonymous locus","NAK-ATPase","NAK-ATPase","PTGS","PTGS","GH7D","GH7A","MnSOD","hemocyanin","CsDll","RNAseq contig","microsat")
)

#initiate vectors
fasta_header=c()
sequence=c()

species_name=c()
organism=c()
accession=c()
description=c()

voucher=c()
country=c()
isolate=c()
gene=c()
bp_length=c()
taxon_id=c()

author_ref1=c()
title_ref1=c()
journal_ref1=c()

author_ref2=c()
title_ref2=c()
journal_ref2=c()

author_ref3=c()
title_ref3=c()
journal_ref3=c()

others=c()

#parse fasta (& genbank) data
seq=""
record=0
for (line in fasta) {
  # if a header is detected:
  if (str_detect(line,fixed(">"))){
    
    #if sequence was recorded for previous record, add it to the sequence vector
    if (nchar(seq)>0){sequence=append(sequence,seq)}
    seq="" #reset seq
    
    record=record+1 #track record number
    
    #store fasta header
    fasta_header = append(fasta_header,line)
    
    #parse fasta header
    header_split=strsplit(line, "|", fixed=TRUE)
    
    #accession from fasta
    accession=append(accession,header_split[[1]][4])
    
    #description from fasta
    des=str_trim(header_split[[1]][5], side="left")
    description=append(description,des)
    
    #species name from fasta
    #[cfa. ]{0,5} added to accomodate "Genus cf. species" and "Genus aff. species"
    species_name=append(species_name,str_extract(des,"[A-Z][a-z]+ [cfa. ]{0,5}[a-z.]+"))
    
    #gene from fasta
    ingenes=FALSE
    gen=""
    for(i in seq(length(genelookup$rawlist))){
      if(str_detect(des,as.character(genelookup$rawlist[i]))){
        lookup=as.character(genelookup$genelist[i])
        if(!str_detect(gen,lookup)){gen=paste(lookup,gen,sep=" ")}
        ingenes=TRUE
      }
    }
    if(!ingenes){
      gen = "other"
      others=append(others,des)
    }
    gene=append(gene,gen)
    
    
    
    #parse data from genbank file:
    gb=genbank_vector[record]
    
    #organism from gb
    org_pattern="organism=\"([^\"]+)\""
    org=str_replace(str_extract(gb,org_pattern),org_pattern,"\\1")
    organism=append(organism,org)
    
    #voucher number from gb
    vou_pattern="voucher=\"([^\"]+)\""
    vou=str_replace(str_extract(gb,vou_pattern),vou_pattern,"\\1")
    voucher=append(voucher,vou)
    
    #country from gb
    country_pat="country=\"([^\"]+)\""
    coun=str_replace(str_extract(gb,country_pat),country_pat,"\\1")
    country=append(country,coun)
    
    #isolate number from gb
    iso_pattern="/isolate=\"([^\"]+)\""
    iso=str_replace(str_extract(gb,iso_pattern),iso_pattern,"\\1")
    isolate=append(isolate,iso)
    
    #sequence length from gb
    bps=as.numeric(str_extract(str_extract(gb,"[:digit:]+ bp"),"[:digit:]+"))
    bp_length=append(bp_length,bps)
    
    #taxon ID from gb
    tax=as.numeric(str_extract(str_extract(gb,"taxon:[:digit:]+"),"[:digit:]+"))
    taxon_id=append(taxon_id,tax)
    
    #author and/or consortium for References 1-3 from gb
    reference_ends=c("REFERENCE","COMMENT","FEATURES")
    for(end in reference_ends){gb=str_replace_all(gb,end,paste("@",end,sep=""))}
    reference_keys=c("AUTHORS","CONSRTM","TITLE","JOURNAL")
    for(key in reference_keys){gb=str_replace_all(gb,key,paste("%",key,sep=""))}
    
    #REFERENCES 1-3
    for (i in 1:3) {
      auth_pat=paste("REFERENCE ","[^@]+%AUTHORS ([^%]+) %",sep=as.character(i))
      cons_pat=paste("REFERENCE ","[^@]+%CONSRTM ([^%]+) %",sep=as.character(i))
      title_pat=paste("REFERENCE ","[^@]+%TITLE ([^%]+) %",sep=as.character(i))
      journ_pat=paste("REFERENCE ","[^@]+%JOURNAL ([^%@]+) [%@]",sep=as.character(i))
      auth= str_replace(str_extract(gb,auth_pat),auth_pat,"\\1")
      cons= str_replace(str_extract(gb,cons_pat),cons_pat,"\\1")
      title= str_replace(str_extract(gb,title_pat),title_pat,"\\1")
      journ= str_replace(str_extract(gb,journ_pat),journ_pat,"\\1")
      if(is.na(auth)){
        auth=cons # in case consortium listed but no author
      }else if(!is.na(cons)){
        auth=paste(auth,cons,sep=", ") # in case author & consortium listed (never?)
      }
      if(i==1){
        author_ref1=append(author_ref1,auth)
        title_ref1=append(title_ref1,title)
        journal_ref1=append(journal_ref1,journ)
      }else if(i==2){
        author_ref2=append(author_ref2,auth)
        title_ref2=append(title_ref2,title)
        journal_ref2=append(journal_ref2,journ)
      }else if(i==3){
        author_ref3=append(author_ref3,auth)
        title_ref3=append(title_ref3,title)
        journal_ref3=append(journal_ref3,journ)
      }
    }
    
  } else {
    # if the line is not a header, add the line to seq
    seq=paste(seq,line,sep="")
  }
}
sequence=append(sequence,seq)

df=data.frame(species_name,organism,taxon_id,voucher,isolate,country,accession,gene,bp_length,author_ref1,title_ref1,journal_ref1,author_ref2,title_ref2,journal_ref2,author_ref3,title_ref3,journal_ref3,description,fasta_header,sequence,stringsAsFactors=FALSE)

print(others)

```

##Build Gene Table

```{r, eval=FALSE, tidy=TRUE}
#initialize gene vectors:
mt_Genome=mt_COI=n_H3=mt_12S=mt_16S=n_18S=n_28S=n_RpS5=n_RpS2=ArgKin=EF1a=GAPDH=ID=MDH=CAD=Wgl=mt_COX2=mt_COX3=mt_ATP6=mt_CytB=HSP70=n_ITS2=rep("",nrow(df))

for(i in seq(nrow(df))){
  entry=as.character(df$gene[i])
  if(str_detect(entry,"mt Genome")){
    mt_Genome[i]=as.character(df$accession[i])
    mt_COI[i]=mt_12S[i]=mt_16S[i]=mt_COX2[i]=mt_COX3[i]=mt_ATP6[i]=mt_CytB[i]="mt Genome"
  }
  if(str_detect(entry,"COI")){mt_COI[i]=as.character(df$accession[i])}
  if(str_detect(entry,"H3")){n_H3[i]=as.character(df$accession[i])}
  if(str_detect(entry,"12S")){mt_12S[i]=as.character(df$accession[i])}
  if(str_detect(entry,"16S")){mt_16S[i]=as.character(df$accession[i])}
  if(str_detect(entry,"18S")){n_18S[i]=as.character(df$accession[i])}
  if(str_detect(entry,"28S")){n_28S[i]=as.character(df$accession[i])}
  if(str_detect(entry,"RpS5")){n_RpS5[i]=as.character(df$accession[i])}
  if(str_detect(entry,"RpS2")){n_RpS2[i]=as.character(df$accession[i])}
  if(str_detect(entry,"ArgKin")){ArgKin[i]=as.character(df$accession[i])}
  if(str_detect(entry,"EF1a")){EF1a[i]=as.character(df$accession[i])}
  if(str_detect(entry,"GAPDH")){GAPDH[i]=as.character(df$accession[i])}
  if(str_detect(entry,"ID")){ID[i]=as.character(df$accession[i])}
  if(str_detect(entry,"MDH")){MDH[i]=as.character(df$accession[i])}
  if(str_detect(entry,"CAD")){CAD[i]=as.character(df$accession[i])}
  if(str_detect(entry,"Wgl")){Wgl[i]=as.character(df$accession[i])}
  if(str_detect(entry,"COX2")){mt_COX2[i]=as.character(df$accession[i])}
  if(str_detect(entry,"COX3")){mt_COX3[i]=as.character(df$accession[i])}
  if(str_detect(entry,"ATP6")){mt_ATP6[i]=as.character(df$accession[i])}
  if(str_detect(entry,"CytB")){mt_CytB[i]=as.character(df$accession[i])}
  if(str_detect(entry,"HSP70")){HSP70[i]=as.character(df$accession[i])}
  if(str_detect(entry,"ITS2")){n_ITS2[i]=as.character(df$accession[i])}
}

gene_table=df[,c(2:4)]

gene_table=data.frame(gene_table,mt_Genome,mt_COI,n_18S,n_28S,n_H3,mt_12S,mt_16S,n_RpS5,n_RpS2,ArgKin,EF1a,GAPDH,ID,MDH,CAD,Wgl,mt_COX2,mt_COX3,mt_ATP6,mt_CytB,HSP70,n_ITS2,stringsAsFactors=FALSE)

#aggregate rows where voucher # matches (omits rows where voucher is NA)
gt=aggregate(gene_table, by=list(gene_table$voucher), max)

#remove Group.1 row
gt=gt[,c(2:ncol(gt))]

#replace rows where voucher is NA
gt=rbind(gt,gene_table[is.na(voucher),])
```

##Output Data

```{r, eval=FALSE, tidy=TRUE}
#parsed data table
df_filename=paste(QUERY,DATABASE,"alldata.txt",sep=".")
write.table(df, df_filename, sep="\t",row.names=FALSE)
#gene table
gt_filename=paste(QUERY,DATABASE,"genetable.txt",sep=".")
write.table(gt, gt_filename, sep="\t",row.names=FALSE)

```

Notes:
Check gene calls for these Accession #s (from Corophiida nucleotide data):
"large subunit ribosomal RNA gene,
            partial sequence; mitochondrial"
KP316286 KP316285 KP316276 KP316274 KP316272 ...
#currently recorded as 16S mitochondrial

Check "small subunit ribosomal"
#currently recorded as 18S nuclear

Lots of RNAseq contigs listed for QUERY="Gammarida"
"contig_\\d+_\\d+ transcribed RNA"
#currently recorded as RNAseq contig

ToDo:
1. Add gene lookup table genes to gene table



